crates/oidc-client/src/utils/client_credentials_layer.rs

```diff
@@ -0,0 +1,714 @@
+
+                let client_assertion = match jwt_signing_method {
+                    JwtSigningMethod::Keystore(keystore) => {
+                        let key = keystore.signing_key_for_algorithm(signing_alg).ok_or(
+                            IntoCredentialsError::InvalidCredentials(
+                                "no suitable key for signing algorithm",
+                            ),
+                        )?;
+                        let signer = key.params().signing_key_for_alg(signing_alg)?;
+                        let header = JsonWebSignatureHeader::new(signing_alg.clone());
+                        Jwt::sign(header, claims, &signer)?.to_string()
+                    }
+                    JwtSigningMethod::Custom(jwt_signing_fn) => jwt_signing_fn(claims).await,
+                };
+
+                Self::Jwt {
+                    client_id,
+                    client_assertion,
+                    client_assertion_type: JwtBearerClientAssertionType,
+                }
+            }
+            _ => return Err(IntoCredentialsError::UnsupportedMethod),
+        };
+
+        Ok(res)
+    }
+}
+
+fn prepare_claims(iss: String, aud: String) -> Result<HashMap<String, Value>, ClaimError> {
+    let mut claims = HashMap::new();
+    let now = Utc::now();
+    claims::ISS.insert(&mut claims, iss.clone())?;
+    claims::SUB.insert(&mut claims, iss)?;
+    claims::AUD.insert(&mut claims, aud)?;
+    claims::IAT.insert(&mut claims, now)?;
+    claims::EXP.insert(&mut claims, now + Duration::minutes(5))?;
+
+    let mut jti = [0u8; 16];
+    rand::thread_rng().fill(&mut jti);
+    let jti = Base64UrlUnpadded::encode_string(&jti);
+    claims::JTI.insert(&mut claims, jti)?;
+
+    Ok(claims)
+}
+
+#[derive(Clone, Serialize)]
+#[skip_serializing_none]
+pub struct ClientCredentialsRequest<T: Serialize> {
+    #[serde(flatten)]
+    body: T,
+    #[serde(flatten)]
+    credentials: Option<ClientCredentials>,
+}
+
+#[derive(Clone)]
+pub(crate) struct ClientCredentialsService<S> {
+    inner: S,
+    params: Arc<ClientCredentialsParams>,
+}
+
+impl<S> ClientCredentialsService<S> {
+    pub(crate) const fn new(inner: S, params: Arc<ClientCredentialsParams>) -> Self {
+        Self { inner, params }
+    }
+}
+
+impl<S, T> Service<Request<T>> for ClientCredentialsService<S>
+where
+    S: Service<Request<ClientCredentialsRequest<T>>> + Clone + Send + 'static,
+    S::Future: Send + 'static,
+    S::Error: 'static,
+    T: Serialize + Send + 'static,
+{
+    type Error = Error<S::Error>;
+    type Response = S::Response;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Error::Service)
+    }
+
+    fn call(&mut self, request: Request<T>) -> Self::Future {
+        let params = self.params.clone();
+        let clone = self.inner.clone();
```

Comment:
I'm not a big fan of having to clone the service, but I couldn't figure out another way to make this method not depend on the lifetime of `self`, because there is a possible future to await before making the request.

Reply:
I'd change up your layer to have the `ClientCredentials` directly in them instead of loading them on-the-fly. This would also avoid having to box the future, since you could directly use `S::Future`, and also probably to `Arc` the params, since you would build the request all synchronously